tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,584 @@
+
+  /**
+   * TiKV recommends each RPC packet should be less than ~1MB. We keep each packet's Key+Value size
+   * below 16KB.
+   */
+  private static final int TXN_COMMIT_BATCH_SIZE = 16 * 1024;
+
+  /** unit is second */
+  private static final long DEFAULT_BATCH_WRITE_LOCK_TTL = 3000;
+
+  private static final Logger LOG = LoggerFactory.getLogger(TiBatchWrite2PCClient.class);
+
+  private TxnKVClient kvClient;
+  private RegionManager regionManager;
+
+  /** start timestamp of transaction which get from PD */
+  private final long startTs;
+
+  public TiBatchWrite2PCClient(TxnKVClient kvClient, long startTime) {
+    this.kvClient = kvClient;
+    this.regionManager = kvClient.getRegionManager();
+    this.startTs = startTime;
+  }
+
+  /**
+   * 2pc - prewrite primary key
+   *
+   * @param backOffer
+   * @param primaryKey
+   * @param value
+   * @return
+   */
+  public void prewritePrimaryKey(BackOffer backOffer, byte[] primaryKey, byte[] value)
+      throws TiBatchWriteException {
+    String error = this.doPrewritePrimaryKey(backOffer, primaryKey, value);
+    if (error != null) {
+      throw new TiBatchWriteException("prewrite primary key error: " + error);
+    }
+  }
+
+  private String doPrewritePrimaryKey(BackOffer backOffer, byte[] key, byte[] value) {
+    TiRegion tiRegion = this.regionManager.getRegionByKey(ByteString.copyFrom(key));
+    BatchKeys batchKeys = new BatchKeys(tiRegion.getId(), Arrays.asList(value));
+
+    Kvrpcpb.Mutation mutation =
+        Kvrpcpb.Mutation.newBuilder()
+            .setKey(ByteString.copyFrom(key))
+            .setValue(ByteString.copyFrom(value))
+            .setOp(Kvrpcpb.Op.Put)
+            .build();
+    List<Kvrpcpb.Mutation> mutationList = Collections.singletonList(mutation);
+
+    // send rpc request to tikv server
+    long regionId = batchKeys.getRegioId();
+    long lockTTL = getTxnLockTTL(this.startTs);
+    ClientRPCResult prewriteResult =
+        this.kvClient.prewrite(backOffer, mutationList, key, lockTTL, this.startTs, regionId);
+    if (!prewriteResult.isSuccess() && !prewriteResult.isRetry()) {
+      return prewriteResult.getError();
+    }
+    if (prewriteResult.isRetry()) {
+      try {
+        backOffer.doBackOff(
+            BackOffFunction.BackOffFuncType.BoRegionMiss,
+            new GrpcException(
+                String.format(
+                    "Txn prewrite primary key failed, regionId=%s, detail=%s",
+                    batchKeys.getRegioId(), prewriteResult.getError())));
+        // re-split keys and commit again.
+        return this.doPrewritePrimaryKey(backOffer, key, value);
+      } catch (GrpcException e) {
+        String error =
+            String.format(
+                "Txn prewrite primary key error, re-split commit failed, regionId=%s, detail=%s",
+                batchKeys.getRegioId(), e.getMessage());
+        LOG.error(error);
+        return error;
+      }
+    }
+    // success return
+    return null;
```

Comment:
null means success is kinda wired to me.

Reply:
me too

Reply:
fix, will throw Exception